### ISSUE TITLE
"All or None" orders

### ISSUE THREAD

**Issue by jkmaina** [viabtc/viabtc_exchange_server#164]
I have a use case where an order should be fully matched or not at all. For example, when you place a sell order of 2 and you only want a match with a buy order of 2 or more (if lets say buy is 2.5 then it should also match also and trade the 2, with the extra 0.5 quantity remaining as a pending for the buyer.)

How can i implement this?

**Comment by objectt** [comment]
I will work on this. What is the term for this type of order? I guess it is similar to "market order" but with the entire order guaranteed to be processed at the same price.

**Comment by objectt** [comment]
@jkmaina 

{"method": "order.put_aon", "params": [1, "BTC",  2, "20", "105", "0", "dev"], "id": 1516681174}
params : [ _user_id, market, [sell/buy], amount, price, taker_fee, source_]

https://github.com/objectt/viabtc_exchange_server/tree/order__aon

I am not sure if this change is appropriate for a pull request. It is still under test, please let me know if you find any issues.

**Comment by jkmaina** [comment]
Thank you! I will test and get back you asap

**Comment by jkmaina** [comment]
Great work @objectt on the AON. Here is my feedback on testing:

AON Orders get posted OK but are not matching correctly . Also I am not getting AON orders in the pending order list.

- When I post an aon sell order(side=1), the order gets posted OK with type 3 response.

`{"method": "order.put_aon", "params": [6,"BCHBTC",1,"5","5291","0","api.v1"],"id":123}`

response:

`{
    "id": 123,
    "error": null,
    "result": {
        "price": "5291",
        "mtime": 1544478597.98857,
        "id": 27,
        "market": "BCHBTC",
        "user": 6,
        "deal_fee": "0",
        "type": 3,
        "source": "api.v1",
        "maker_fee": "0",
        "side": 1,
        "ctime": 1544478597.98857,
        "deal_stock": "0",
        "amount": "5",
        "left": "5",
        "taker_fee": "0",
        "deal_money": "0"
    }
}`

However, when i post a buy order (side=2), for same quantity and same price, it does not match - deal stock/money remains zero.

`{"method": "order.put_aon", "params": [3,"BCHBTC",1,"5","5291","0","api.v1"],"id":123}`

response:

'{
    "id": 123,
    "error": null,
    "result": {
        "price": "5291",
        "mtime": 1544478961.5650671,
        "id": 30,
        "market": "BCHBTC",
        "user": 3,
        "deal_fee": "0",
        "type": 3,
        "source": "api.v1",
        "maker_fee": "0",
        "side": 2,
        "ctime": 1544478961.5650671,
        "deal_stock": "0",
        "amount": "5",
        "left": "5",
        "taker_fee": "0",
        "deal_money": "0"
    }
}

Also, when i check for pending orders, they don't show AON orders. Other order types still show.

`{"method": "order.pending", "params": [3,"BCHBTC",0,100],"id":123}`
response:
`{
    "id": 123,
    "error": null,
    "result": {
        "limit": 100,
        "offset": 0,
        "total": 0,
        "records": []
    }
}`

I haven't checked finished orders and order deals because of the first issue on matching, but could need validation also.

Could you kindly have a look and update?

Many Thanks
James

**Comment by objectt** [comment]
@jkmaina Thanks for sparing time to test it. 

Just like a market order,  AON order is not created if it is not matched immediately (Meaning there is only taker for AON orders; no makers). However, this can be adjusted if needed. 

In the mean time, place some random limit orders first then try creating AON orders.

**Comment by objectt** [comment]
@jkmaina Sounds good! Will work on it right away.

**Comment by jkmaina** [comment]
i agree. cancellation is a manual process from the UI (or automated externally). I think you don't really need to factor that in at all - this is taken care of with the order.cancel api already in place. 

The difference between _aon_ and _fill or kill_ is the behaviour - the aon order is more like a limit order, it stays on the order book and can get matched either side of a buy/sell of an opposite aon or limit order or market order which fill it completely. The pending order stays until the exchange decides what to do with it via UI or other external process such as session management. 

While the _fill or kill_ is executed immediately like a market order as long as its fill order conditions are met. This is exactly what you have implemented.
